### PR TITLE
feat: implement decrease_position_collateral_utils

### PR DESCRIPTION
**Pull request by zarboq**
# Pull Request type

- Feature
- Testing
Resolves: #328 

Implemented `decrease_position_collateral_utils` and basic tests.

### CODE REVIEW

**Review by zarboq — COMMENTED** [review]
I'll fix failing tests

File: src/market/market_utils.cairo [hunk 1]
```diff
@@ -497,7 +497,6 @@ fn get_max_open_interest(
 /// * `delta` - The amount to increment by.
 fn increment_claimable_collateral_amount(
     data_store: IDataStoreDispatcher,
-    chain: IChainDispatcher,
     event_emitter: IEventEmitterDispatcher,
     market_address: ContractAddress,
     token: ContractAddress,
```
Comment (Sk8erboi84) [inline src/market/market_utils.cairo:502]: If you wanna go with the `market_address` change the other to `XXX_address` or do the other way. Do that for the rest of the functions too please

**Review by Sk8erboi84 — CHANGES_REQUESTED** [review]

File: src/position/decrease_position_collateral_utils.cairo [hunk 2]
```diff
@@ -43,86 +42,491 @@ struct PayForCostResult {
...
+        .position
+        .size_in_usd
+        && (base_order_utils::is_liquidation_order(params.order.order_type)
+            || params.secondary_order_type == order::SecondaryOrderType::Adl(()));
+    // in case price impact is too high it is capped and the difference is made to be claimable
+    // the execution price is based on the capped price impact so it may be a better price than what it should be
+    // price_impact_diff_usd is the difference between the maximum price impact and the originally calculated price impact
+    // e.g. if the originally calculated price impact is -$100, but the capped price impact is -$80
+    // then priceImpactDiffUsd would be $20
+    let (price_impact_usd_, price_impact_diff_usd_, execution_price_) = get_execution_price(
```
Comment (Sk8erboi84) [inline src/position/decrease_position_collateral_utils.cairo:92]: dont put _ at the end try to find better naming (for all)

File: src/position/decrease_position_collateral_utils.cairo [hunk 3]
```diff
@@ -43,86 +42,491 @@ struct PayForCostResult {
...
-        position_fee_amount: 0,
-        total_cost_amount_excluding_funding: 0,
-        total_cost_amount: 0,
+
+    let mut fees: position_pricing_utils::PositionFees = position_pricing_utils::get_position_fees(
+        get_position_fees_params
+    );
+
+    // if the pnl is positive, deduct the pnl amount from the pool
+    if values.base_pnl_usd > 0 {
```
Comment (Sk8erboi84) [inline src/position/decrease_position_collateral_utils.cairo:133]: add this  `// use pnlTokenPrice.max to minimize the tokens paid out` after the if statement

**Review by Sk8erboi84 — CHANGES_REQUESTED** [review]
Maybe you can store the strings : "impact" , "fees"... at the top of the file or in another file and call them

File: src/position/decrease_position_collateral_utils.cairo [hunk 4]
```diff
@@ -43,86 +42,491 @@ struct PayForCostResult {
...
+
+    // the priceImpactDiffUsd has been deducted from the output amount or the position's collateral
+    // to reduce the chance that the position's collateral is reduced by an unexpected amount, adjust the
+    // initialCollateralDeltaAmount by the priceImpactDiffAmount
+    // this would also help to prevent the position's leverage from being unexpectedly increased
+    //
+    // note that this calculation may not be entirely accurate since it is possible that the priceImpactDiffUsd
+    // could have been paid with one of or a combination of collateral / outputAmount / secondaryOutputAmount
+    if params.order.initial_collateral_delta_amount > 0 && values.price_impact_diff_usd > 0 {
+        let initial_collateral_delta_amount = params.order.initial_collateral_delta_amount;
```
Comment (Sk8erboi84) [inline src/position/decrease_position_collateral_utils.cairo:489]: set the type of each variable please

File: src/position/decrease_position_collateral_utils.cairo [hunk 5]
```diff
@@ -43,86 +42,491 @@ struct PayForCostResult {
...
+                params.order_key,
+                params.order.initial_collateral_delta_amount, // collateral_delta_amount
+                values.remaining_collateral_amount // next_collateral_delta_amount
+            );
+
+        params.order.initial_collateral_delta_amount = values.remaining_collateral_amount;
+    }
+
+    if params.order.initial_collateral_delta_amount > 0 {
+        values.remaining_collateral_amount = params.order.initial_collateral_delta_amount;
```
Comment (Sk8erboi84) [inline src/position/decrease_position_collateral_utils.cairo:525]: You should put `-=`
```suggestion
        values.remaining_collateral_amount -= params.order.initial_collateral_delta_amount;
```

**Review by Sk8erboi84 — CHANGES_REQUESTED** [review]

File: src/position/decrease_position_collateral_utils.cairo [hunk 6]
```diff
@@ -43,86 +42,491 @@ struct PayForCostResult {
...
-    let position_borrowing_fees = PositionBorrowingFees {
-        borrowing_fee_usd: 0,
-        borrowing_fee_amount: 0,
-        borrowing_fee_receiver_factor: 0,
-        borrowing_fee_amount_for_fee_receiver: 0,
-    };
-    let position_ui_fees = PositionUiFees {
-        ui_fee_receiver: address_zero, ui_fee_receiver_factor: 0, ui_fee_amount: 0,
+    mut params: position_utils::UpdatePositionParams, cache: position_utils::DecreasePositionCache
+) -> (position_utils::DecreasePositionCollateralValues, position_pricing_utils::PositionFees) {
```
Comment (Sk8erboi84) [inline src/position/decrease_position_collateral_utils.cairo:61]: Do you need to put the full path for the struct ? 
```position_utils::UpdatePositionParams``` cant you just put ```UpdatePositionParams```

File: src/position/decrease_position_collateral_utils.cairo [hunk 7]
```diff
@@ -186,121 +708,83 @@ fn pay_for_cost(
...
+fn get_empty_fees(
+    fees: @position_pricing_utils::PositionFees
+) -> position_pricing_utils::PositionFees {
+    let referral: position_pricing_utils::PositionReferralFees = Default::default();
+
+    // allow the accumulated funding fees to still be claimable
+    // return the latestFundingFeeAmountPerSize, latest_long_token_claimable_funding_amount_per_size,
+    // latest_short_token_claimable_funding_amount_per_size values as these may be used to update the
+    // position's values if the position will be partially closed
+    let funding = position_pricing_utils::PositionFundingFees {
```
Comment (Sk8erboi84) [inline src/position/decrease_position_collateral_utils.cairo:766]: Try adding the type for that variable and all other variable where you did set the type

File: src/position/decrease_position_collateral_utils.cairo [hunk 8]
```diff
@@ -186,121 +708,83 @@ fn pay_for_cost(
...
-        collateral_token_price: price,
+    let borrowing: position_pricing_utils::PositionBorrowingFees = Default::default();
+    let ui: position_pricing_utils::PositionUiFees = Default::default();
+    // all fees are zeroed even though funding may have been paid
+    // the funding fee amount value may not be accurate in the events due to this
+    position_pricing_utils::PositionFees {
+        referral: referral,
+        funding: funding,
+        borrowing: borrowing,
+        ui: ui,
```
Comment (Sk8erboi84) [inline src/position/decrease_position_collateral_utils.cairo:786]: You can just put the name instantly cause they have the same name
```suggestion
        referral,
        funding,
        borrowing,
        ui,
```

**Review by Sk8erboi84 — CHANGES_REQUESTED** [review]

File: src/position/decrease_position_collateral_utils.cairo (hunk 5 above)
Comment (zarboq) [reply src/position/decrease_position_collateral_utils.cairo:525]: Good catch

File: src/position/decrease_position_collateral_utils.cairo (hunk 7 above)
Comment (zarboq) [reply src/position/decrease_position_collateral_utils.cairo:766]: I didn't put it here bc we see the type next to it

Comment (Sk8erboi84) [reply src/position/decrease_position_collateral_utils.cairo:767]: Yes put I think its better to be explicit, we have nothing to lose

**Review by Sk8erboi84 — APPROVED** [review]

**Pull request merged**